util/src/main/java/io/camunda/zeebe/util/FatalErrorHandler.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.util;
+
+import java.lang.Thread.UncaughtExceptionHandler;
+import org.slf4j.Logger;
+
+public final class FatalErrorHandler implements UncaughtExceptionHandler {
+  private final Logger log;
+
+  public FatalErrorHandler(final Logger log) {
+    this.log = log;
+  }
+
+  public void handleError(final Throwable e) {
+    if (e instanceof VirtualMachineError) {
+      log.error("An unrecoverable VM error was thrown, exiting now.", e);
+      System.exit(1);
```

Comment:
:wrench: Maybe we want to use a specific error code to indicate OOM? 137 is the one used by the Linux OOM killer (or 128 + 9). We could reuse this, or we could add one. My thoughts here are that when a user specifies "hey the program exited with X", then we know for sure it's because of this. I know we have the log statement of course, but having a specific status code can be useful if you want to react automatically to it. Let me know what you think.

Reply:
I like to idea to use a custom error code. I'd prefer not to use `137` for OOM precisely because it is so commonly used. I believe that this would make it a bit more difficult to distinguish between external OOM handling (the linux OOM killer, kubernetes etc.) and internal, JVM OOM handling. I'd suggest using exit code 156, which is ascii Z + B and as far as I can see only used by one product that isn't easily confused with Zeebe.